text/0000-promise.md

```diff
@@ -0,0 +1,260 @@
+}
+
+impl JsPromise {
+    pub(crate) fn new_internal<'a>(value: raw::Local) -> Handle<'a, JsPromise> {
+        Handle::new_internal(JsPromise(value))
+    }
+}
+```
+
+## `Deferred`
+
+`Send` handle for resolving or rejecting a `JsPromise`.
+
+```rust
+pub struct Deferred(*mut c_void);
+
+unsafe impl Send for Deferred {}
+
+impl Deferred {
+    /// Resolves a deferred Promise with the value
+    fn resolve<'a, C: Context<'a>, T: Value>(self, cx: &mut C, v: Handle<T>);
+
+    /// Rejects a deferred Promise with the error
+    fn reject<'a, C: Context<'a>, E: Value>(self, cx: &mut C, err: Handle<E>);
+
+    /// Resolves or rejects a deferred Promise with the contents of a `Result`
+    fn complete<'a, C: Context<'a>, T: Value, E: Value>(self, cx: &mut C, result: Result<T, E>);
+}
+```
+
+Similar to [`Persistent`](https://github.com/neon-bindings/rfcs/pull/32), if a `Deferred` is not resolved or rejected, the `Promise` will leak. To help catch this and guide users towards a correct implementation, `Deferred` should `panic` on `Drop` if not used.
+
+```rust
+impl std::ops::Drop for Deferred {
+    fn drop(&mut self) {
+        panic!("JsPromise leaked. Deferred must be used.");
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None? :grin:
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+## High Level Promise Tasks
+
+Using `JsPromise` along with other async features requires careful and verbose usage of several features (`JsPromise`, `EventQueue` and `try_catch`). Neon could exclusively provide a high-level API similar to `Task` or the proposed `TaskBuilder`.
+
+```rust
+fn async_task(cx: FunctionContext) -> JsResult<JsPromise> {
+    let promise = cx.task(|| /* perform async task */)
+        .complete(|cx| /* convert to js types */);
+
+    Ok(promise)
+}
+```
+
+This API is very ergonomic, but removes a considerable amount of flexibility and power from the user. Instead, the initial implementation will focus on orthogonal core primitives like `Persistent`, `EventQueue`, `JsBox` and `JsPromise` that can later be combined into high-level APIs.
+
+## Common `Deferred` pattern
+
+Most promise libraries that provide a `Deferred` object provide the `promise` as part of that object. Neon might have a similar approach:
+
+```rust
+struct Deferred<'a> {
+    handle: *mut c_void,
+    promise: Handle<'a, JsPromise>,
+}
+```
+
+The issue with this approach is that `Deferred` is `Send` and `JsPromise` is *not*. We would also need to provide a getter for the resolve/reject handle. That getter would need to _consume_ self because it cannot be used multiple times. This would result in worse ergonomics.
+
+Instead a tuple is returned, similar to `(sender, receiver)` returned from `std::sync::mpsc::channel`.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should there be a constructor method on `JsPromise`? This is slightly awkward because it's impossible to create a `Promise` without also creating a `Deferred`.
```

Comment:
Yeah it seems OK to me to only have it on the context. We could also have `neon::types::promise(&mut cx)` I suppose, but it's really not providing any value over just `cx.promise()`.

Reply:
For symmetry, I think there should be a constructor on the type. In this example, `Deferred` is a placeholder for the final name:

```rust
// All equivalent
cx.promise()
JsPromise::deferred(&mut cx)
Deferred::new(&mut cx)
```

Reply:
The symmetry should not be included. We can take a look at the `mpsc` API where there is no way to detect a `Sender` or `Receiver`. They are only created with `channel()`.

Similarly, the only way to create `JsPromise` and `Deferred` is with `cx.promise()`.